backend/groth16/bls12-377/goicicle_wrapper.go

```diff
@@ -0,0 +1,250 @@
+	var timings []time.Duration
+	evalTime := time.Now()
+	res := icicle.Evaluate(scalars_out, scalars_d, twiddles_d, coset_powers_d, size, twid_size, isCoset)
+	evalTimeElapsed := time.Since(evalTime)
+	timings = append(timings, evalTimeElapsed)
+
+	if res != 0 {
+		fmt.Print("Issue evaluating")
+	}
+
+	revTime := time.Now()
+	icicle.ReverseScalars(scalars_out, size)
+	revTimeElapsed := time.Since(revTime)
+	timings = append(timings, revTimeElapsed)
+
+	return timings
+}
+
+func MsmBls12377GnarkAdapter(points []curve.G1Affine, scalars []fr.Element) (curve.G1Jac, error, []time.Duration) {
+	var timings []time.Duration
+	out := new(icicle.G1ProjectivePoint)
+
+	convSTime := time.Now()
+	parsedScalars := bls12377.BatchConvertFromFrGnark(scalars)
+	timings = append(timings, time.Since(convSTime))
+
+	convPTime := time.Now()
+	parsedPoints := bls12377.BatchConvertFromG1Affine(points)
+	timings = append(timings, time.Since(convPTime))
+
+	msmTime := time.Now()
+	_, err := icicle.Msm(out, parsedPoints, parsedScalars, 0)
+	timings = append(timings, time.Since(msmTime))
+
+	return *bls12377.G1ProjectivePointToGnarkJac(out), err, timings
+}
+
+func PolyOps(a_d, b_d, c_d, den_d unsafe.Pointer, size int) (timings []time.Duration) {
+	convSTime := time.Now()
+	ret := icicle.VecScalarMulMod(a_d, b_d, size)
+	timings = append(timings, time.Since(convSTime))
+
+	if ret != 0 {
+		fmt.Print("Vector mult a*b issue")
+	}
+	convSTime = time.Now()
+	ret = icicle.VecScalarSub(a_d, c_d, size)
+	timings = append(timings, time.Since(convSTime))
+
+	if ret != 0 {
+		fmt.Print("Vector sub issue")
+	}
+	convSTime = time.Now()
+	ret = icicle.VecScalarMulMod(a_d, den_d, size)
+	timings = append(timings, time.Since(convSTime))
+
+	if ret != 0 {
+		fmt.Print("Vector mult a*den issue")
+	}
+
+	return
+}
+
+func MsmOnDevice(scalars_d, points_d unsafe.Pointer, count, bucketFactor int, convert bool) (curve.G1Jac, unsafe.Pointer, error, time.Duration) {
+	out_d, _ := cudawrapper.CudaMalloc(96)
+
+	msmTime := time.Now()
+	icicle.Commit(out_d, scalars_d, points_d, count, bucketFactor)
+	timings := time.Since(msmTime)
+
+	if convert {
+		outHost := make([]icicle.G1ProjectivePoint, 1)
+		cudawrapper.CudaMemCpyDtoH[icicle.G1ProjectivePoint](outHost, out_d, 96)
+		return *bls12377.G1ProjectivePointToGnarkJac(&outHost[0]), nil, nil, timings
+	}
+
+	return curve.G1Jac{}, out_d, nil, timings
+}
+
+func MsmG2OnDevice(scalars_d, points_d unsafe.Pointer, count, bucketFactor int, convert bool) (curve.G2Jac, unsafe.Pointer, error, time.Duration) {
+	out_d, _ := cudawrapper.CudaMalloc(192)
+
+	msmTime := time.Now()
+	icicle.CommitG2(out_d, scalars_d, points_d, count, bucketFactor)
+	timings := time.Since(msmTime)
+
+	if convert {
+		outHost := make([]icicle.G2Point, 1)
+		cudawrapper.CudaMemCpyDtoH[icicle.G2Point](outHost, out_d, 192)
+		return *bls12377.G2PointToGnarkJac(&outHost[0]), nil, nil, timings
+	}
```

Comment:
```suggestion
    g2ProjPointBytes = fp.Bytes * 6 // X,Y,Z each with A0, A1 of fp.Bytes
	out_d, _ := cudawrapper.CudaMalloc(g2ProjPointBytes)

	msmTime := time.Now()
	icicle.CommitG2(out_d, scalars_d, points_d, count, bucketFactor)
	timings := time.Since(msmTime)

	if convert {
		outHost := make([]icicle.G2Point, 1)
		cudawrapper.CudaMemCpyDtoH[icicle.G2Point](outHost, out_d, g2ProjPointBytes)
		return *bls12377.G2PointToGnarkJac(&outHost[0]), nil, nil, timings
	}
```

We should change this for BN254 as well so future copied curves aren't following BN254